pkg/controller/assign/assign_controller.go

```diff
@@ -0,0 +1,141 @@
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	logf "sigs.k8s.io/controller-runtime/pkg/log"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+)
+
+var (
+	log = logf.Log.WithName("controller").WithValues(logging.Process, "assignment_controller")
+)
+
+type Adder struct {
+	MutationCache *mutation.System
+}
+
+// Add creates a new Assign Controller and adds it to the Manager. The Manager will set fields on the Controller
+// and Start it when the Manager is Started.
+func (a *Adder) Add(mgr manager.Manager) error {
+	r := newReconciler(mgr, a.MutationCache)
+	return add(mgr, r)
+}
+
+func (a *Adder) InjectOpa(o *opa.Client) {}
+
+func (a *Adder) InjectWatchManager(w *watch.Manager) {}
+
+func (a *Adder) InjectControllerSwitch(cs *watch.ControllerSwitch) {}
+
+func (a *Adder) InjectTracker(t *readiness.Tracker) {}
+
+func (a *Adder) InjectMutationCache(mutationCache *mutation.System) {
+	a.MutationCache = mutationCache
+}
+
+// newReconciler returns a new reconcile.Reconciler
+func newReconciler(mgr manager.Manager, mutationCache *mutation.System) *AssignReconciler {
+	r := &AssignReconciler{system: mutationCache, Client: mgr.GetClient()}
+	return r
+}
+
+// add adds a new Controller to mgr with r as the reconcile.Reconciler
+func add(mgr manager.Manager, r reconcile.Reconciler) error {
+	if !*mutation.MutationEnabled {
+		return nil
+	}
+
+	// Create a new controller
+	c, err := controller.New("assign-controller", mgr, controller.Options{Reconciler: r})
+	if err != nil {
+		return err
+	}
+
+	// Watch for changes to Assign
+	if err = c.Watch(
+		&source.Kind{Type: &mutationsv1alpha1.Assign{}},
+		&handler.EnqueueRequestForObject{}); err != nil {
+		return err
+	}
+	return nil
+}
+
+// AssignReconciler reconciles a Assign object
+type AssignReconciler struct {
+	client.Client
+	system *mutation.System
+}
+
+// +kubebuilder:rbac:groups=mutations.gatekeeper.sh,resources=*,verbs=get;list;watch;create;update;patch;delete
+
+// Reconcile reads that state of the cluster for a Assign object and makes changes based on the state read
+// and what is in the Assign.Spec
+func (r *AssignReconciler) Reconcile(request reconcile.Request) (reconcile.Result, error) {
+	log.Info("Reconcile", "request", request)
+	deleted := false
+	assign := &mutationsv1alpha1.Assign{}
+	err := r.Get(context.TODO(), request.NamespacedName, assign)
+	if err != nil {
+		if !errors.IsNotFound(err) {
+			return reconcile.Result{}, err
+		}
+		deleted = true
+		assign = &mutationsv1alpha1.Assign{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      request.NamespacedName.Name,
+				Namespace: request.NamespacedName.Namespace,
+			},
+			TypeMeta: metav1.TypeMeta{
+				Kind:       "Assign",
+				APIVersion: "mutations.gatekeeper.sh",
+			},
+		}
+	}
+	deleted = deleted || !assign.GetDeletionTimestamp().IsZero()
+
+	mutator, err := mutation.MutatorForAssign(assign)
```

Comment:
I think this will suffer from the same issue @sozercan raised in https://github.com/open-policy-agent/gatekeeper/pull/1013

I opened https://github.com/open-policy-agent/gatekeeper/pull/1020 , and this pr should probably follow once the ID remove logic gets in

Reply:
sg!

Reply:
@fedepaol PTAL